flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/catalog/JDBCCatalog.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.java.io.jdbc.catalog;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.java.io.jdbc.dialect.JDBCDialect;
+import org.apache.flink.api.java.io.jdbc.dialect.JDBCDialects;
+import org.apache.flink.table.catalog.Catalog;
+import org.apache.flink.table.catalog.CatalogBaseTable;
+import org.apache.flink.table.catalog.CatalogDatabase;
+import org.apache.flink.table.catalog.ObjectPath;
+import org.apache.flink.table.catalog.exceptions.CatalogException;
+import org.apache.flink.table.catalog.exceptions.DatabaseNotExistException;
+import org.apache.flink.table.catalog.exceptions.TableNotExistException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+/**
+ *
+ */
+@PublicEvolving
+public class JDBCCatalog extends AbstractJDBCCatalog {
+
+	private static final Logger LOG = LoggerFactory.getLogger(JDBCCatalog.class);
+
+	private final Catalog internal;
+
+	public JDBCCatalog(String catalogName, String defaultDatabase, String username, String pwd, String baseUrl) {
+		super(catalogName, defaultDatabase, username, pwd, baseUrl);
+
+		JDBCDialect dialect = JDBCDialects.get(baseUrl).get();
+
+		if (dialect instanceof JDBCDialects.PostgresDialect) {
+			internal = new PostgresCatalog(catalogName, defaultDatabase, username, pwd, baseUrl);
```

Comment:
A internal Catalog looks weird to me...
I think this is just internal implementation.
Why not introduce something like `JDBCDialect`? PG implementation just need see what they can see.

Reply:
Unlike just dialect, the impl would vary a lot, e.g some db may reuse connections, some (like pg) can't. flink-jdbc is trying to give users a unified interface, thus as you said, it's internal impl and would be ok.

besides, this design is voted and approved in the FLIP

Reply:
I know this implementation is in FLIP.
But this if else still looks weird to me. Can we add this interface to JDBCDialect? like add method: `createCatalog(...)` in JDBCDialect. Instead need update JDBCCatalog every time.

Reply:
I see what you mean now, I'll move to a util class.